### ISSUE TITLE
FRR and MPLS module

### ISSUE THREAD

**Issue by barajus** [ipspace/netlab-examples#23]
Hi back, I am still playing around with FRR and the cool kid (clab).

I was giving a try to https://github.com/ipspace/netlab-examples/tree/master/MPLS/vpn-simple and I found some issues with PE:

```
TASK [run /tmp/config.sh to deploy vrf config from /usr/local/lib/python3.10/dist-packages/netsim/ansible/tasks/../templates/vrf/frr.j2] *****************************************
fatal: [pe1]: FAILED! => changed=true
  cmd:
  - bash
  - /tmp/config.sh
  delta: '0:00:00.008140'
  end: '2024-03-13 21:58:35.758576'
  msg: non-zero return code
  rc: 2
  start: '2024-03-13 21:58:35.750436'
  stderr: 'Error: Unknown device type.'
  stderr_lines: <omitted>
  stdout: ''
  stdout_lines: <omitted>
fatal: [pe2]: FAILED! => changed=true
  cmd:
  - bash
  - /tmp/config.sh
  delta: '0:00:00.006389'
  end: '2024-03-13 21:58:35.813947'
  msg: non-zero return code
  rc: 2
  start: '2024-03-13 21:58:35.807558'
  stderr: 'Error: Unknown device type.'
  stderr_lines: <omitted>
  stdout: ''
  stdout_lines: <omitted>
```

(netlab version 1.7.0)

topology if needed:

```
provider: clab
defaults.device: frr

vrfs:
  red:
  blue:

bgp.as: 65000
mpls.ldp: True
mpls.vpn: True

nodes:
  pe1:
    module: [ vrf,ospf,bgp,mpls ]
  pe2:
    module: [ vrf,ospf,bgp,mpls ]
  h1:
    device: linux
  h2:
    device: linux
  h3:
    device: linux
  h4:
    device: linux

links:
- pe1: { vrf: red }
  h1:
- pe2: { vrf: red }
  h2:
- pe1: { vrf: blue }
  h3:
- pe2: { vrf: blue }
  h4:
- pe1-pe2
```

**Comment by ipspace** [comment]
Unfortunately, we're in the "it works for me" territory (apart from an Ansible SNAFU fixed in https://github.com/ipspace/netlab/commit/801da692c29915851e12966c7709c515ba9f0f57 that obviously does not make your version of Ansible unhappy).

Run "netlab initial -m mpls -vv -l pe1" to get the STDOUT as well (we need to see where the config script crashes).

Also, as containers use the host networking stack, please include the printouts from **uname -a** and **lsb_release -a** I'm running Ubuntu 22.04:

```
$ uname -a
Linux brick2 5.15.0-94-generic #104-Ubuntu SMP Tue Jan 9 15:25:40 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.3 LTS
Release:	22.04
Codename:	jammy
```

Finally, check that the MPLS kernel modules are loaded. This is the printout I'm getting:

```
$ grep mpls /proc/modules
mpls_gso 16384 0 - Live 0x0000000000000000
mpls_iptunnel 20480 4 - Live 0x0000000000000000
mpls_router 45056 1 mpls_iptunnel, Live 0x0000000000000000
ip_tunnel 32768 1 mpls_router, Live 0x0000000000000000
```

If they're not loaded, try `sudo modprobe mpls-router mpls-iptunnel` That command is executed as part of the FRR MPLS configuration, so you should see the modules.

**Comment by barajus** [comment]
Thanks for the fast answer as always - here are the traces

```
ubuntu@primary:~/mpls$ netlab initial -m mpls -vv -l pe1
Ansible playbook args: ['-t', 'module', '-e', 'modlist=mpls', '-vv', '-l', 'pe1']
ansible-playbook [core 2.16.2]
  config file = /home/ubuntu/mpls/ansible.cfg
  configured module search path = ['/home/ubuntu/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.10/dist-packages/ansible
  ansible collection location = /home/ubuntu/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible-playbook
  python version = 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
Using /home/ubuntu/mpls/ansible.cfg as config file
redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
redirecting (type: callback) ansible.builtin.yaml to community.general.yaml
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: initial-config.ansible *************************************************************************************************************************************************
3 plays in /usr/local/lib/python3.10/dist-packages/netsim/ansible/initial-config.ansible
[WARNING]: Could not match supplied host pattern, ignoring: unprovisioned

PLAY [Deploy initial device configuration] ***************************************************************************************************************************************

TASK [set_fact] ******************************************************************************************************************************************************************
task path: /usr/local/lib/python3.10/dist-packages/netsim/ansible/initial-config.ansible:8
redirecting (type: connection) ansible.builtin.docker to community.docker.docker
ok: [pe1] => changed=false
  ansible_facts:
    netlab_device_type: frr
    search_path:
    - /home/ubuntu/mpls/templates
    - ~/.netlab/templates
    - ../templates

TASK [set_fact] ******************************************************************************************************************************************************************
task path: /usr/local/lib/python3.10/dist-packages/netsim/ansible/initial-config.ansible:18
redirecting (type: connection) ansible.builtin.docker to community.docker.docker
ok: [pe1] => changed=false
  ansible_facts:
    ready_script: ''

TASK [Waiting for device to become ready] ****************************************************************************************************************************************
task path: /usr/local/lib/python3.10/dist-packages/netsim/ansible/initial-config.ansible:30
skipping: [pe1] => changed=false
  false_condition: ready_script|default('')
  skip_reason: Conditional result was False

PLAY [Deploy module-specific configurations] *************************************************************************************************************************************

TASK [set_fact] ******************************************************************************************************************************************************************
task path: /usr/local/lib/python3.10/dist-packages/netsim/ansible/initial-config.ansible:77
redirecting (type: connection) ansible.builtin.docker to community.docker.docker
ok: [pe1] => changed=false
  ansible_facts:
    modlist:
    - mpls

TASK [include_tasks] *************************************************************************************************************************************************************
task path: /usr/local/lib/python3.10/dist-packages/netsim/ansible/initial-config.ansible:79
skipping: [pe1] => changed=false
  skipped_reason: No items in the list

PLAY [Deploy custom deployment templates] ****************************************************************************************************************************************
skipping: no hosts matched

PLAY RECAP ***********************************************************************************************************************************************************************
pe1                        : ok=3    changed=0    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0
```

```
ubuntu@primary:~/mpls$ uname -a
Linux primary 5.15.0-97-generic #107-Ubuntu SMP Wed Feb 7 13:26:48 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
ubuntu@primary:~/mpls$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.3 LTS
Release:	22.04
Codename:	jammy
```

```
ubuntu@primary:~/mpls$ sudo modprobe mpls-router mpls-iptunnel
modprobe: FATAL: Module mpls-router not found in directory /lib/modules/5.15.0-97-generic
```

(same result for the VRF module fwiw)

**Comment by barajus** [comment]
Solved the kernel module issue with a kernel upgrade `sudo apt install linux-generic`.
After this, I run this version
```
ubuntu@primary:~$ uname -a
Linux primary 5.15.0-100-generic #110-Ubuntu SMP Wed Feb 7 13:27:48 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

(even though, "Ubuntu 20.04 ships with kernel 5.4; MPLS modules are present by default")

**Comment by barajus** [comment]
Trace updated with the MPLS module loaded:

```
TASK [Load MPLS kernel modules] **********************************************************************************************************************************
changed: [pe1 -> localhost]

TASK [template] **************************************************************************************************************************************************
changed: [pe2]
changed: [pe1]

TASK [set_fact] **************************************************************************************************************************************************
ok: [pe1]
ok: [pe2]

TASK [run /tmp/config.sh to deploy mpls config from /usr/local/lib/python3.10/dist-packages/netsim/ansible/tasks/../templates/mpls/frr.j2] ***********************
fatal: [pe1]: FAILED! => changed=true
  cmd:
  - bash
  - /tmp/config.sh
  delta: '0:00:00.172372'
  end: '2024-03-14 22:30:54.410732'
  msg: non-zero return code
  rc: 13
  start: '2024-03-14 22:30:54.238360'
  stderr: |-
    [359|zebra] done
    [362|ospfd] done
    [363|ospf6d] done
    [358|mgmtd] done
    [364|ldpd] done
    [373|watchfrr] done
    [375|staticd] done
    line 17: Failure to communicate[13] to bgpd, line:   neighbor 10.0.0.2 activate

    line 18: Failure to communicate[13] to bgpd, line:   neighbor 10.0.0.2 send-community both

    [365|bgpd] Configuration file[/etc/frr/frr.conf] processing failure: 13
  stderr_lines: <omitted>
  stdout: |-
    net.mpls.platform_labels = 1048575
    net.mpls.conf.eth3.input = 1
    [362|ospfd] sending configuration
    [359|zebra] sending configuration
    [358|mgmtd] sending configuration
    [363|ospf6d] sending configuration
    [364|ldpd] sending configuration
    [373|watchfrr] sending configuration
    Waiting for children to finish applying config...
    [375|staticd] sending configuration
    [365|bgpd] sending configuration
    % Specify remote-as or peer-group commands first
    % Specify remote-as or peer-group commands first
  stdout_lines: <omitted>
fatal: [pe2]: FAILED! => changed=true
  cmd:
  - bash
  - /tmp/config.sh
  delta: '0:00:00.249634'
  end: '2024-03-14 22:30:54.539529'
  msg: non-zero return code
  rc: 13
  start: '2024-03-14 22:30:54.289895'
  stderr: |-
    [359|ospf6d] done
    [354|mgmtd] done
    [355|zebra] done
    line 17: Failure to communicate[13] to bgpd, line:   neighbor 10.0.0.1 activate

    line 18: Failure to communicate[13] to bgpd, line:   neighbor 10.0.0.1 send-community both

    [361|bgpd] Configuration file[/etc/frr/frr.conf] processing failure: 13
    [369|watchfrr] done
    [360|ldpd] done
    [358|ospfd] done
    [371|staticd] done
  stderr_lines: <omitted>
  stdout: |-
    net.mpls.platform_labels = 1048575
    net.mpls.conf.eth3.input = 1
    Waiting for children to finish applying config...
    [355|zebra] sending configuration
    [358|ospfd] sending configuration
    [359|ospf6d] sending configuration
    [360|ldpd] sending configuration
    [369|watchfrr] sending configuration
    [371|staticd] sending configuration
    [361|bgpd] sending configuration
    [354|mgmtd] sending configuration
    % Specify remote-as or peer-group commands first
    % Specify remote-as or peer-group commands first
  stdout_lines: <omitted>
```

**Comment by ipspace** [comment]
I think you solved the crux of the problem (mpls kernel modules), will add a line to caveats.md.

The last errors are due to BGP neighbors not being configured (BGP module has to be run before the MPLS module). If that's a result of "netlab up", then you hit this: https://blog.ipspace.net/2024/02/ansible-set-operations.html. Upgrading to netlab 1.8.0 should solve that.

**Comment by barajus** [comment]
I think we are getting close, I still hit another little one (with 1.8.0):

```
TASK [Find configuration deployment deploy_script for mpls] ****************************************
ok: [pe1]
ok: [pe2]

TASK [Deploy mpls configuration] *******************************************************************
ERROR! Error when evaluating variable in dynamic parent include path: {{ deploy_script }}. When using static imports, the parent dynamic include cannot utilize host facts or variables from inventory
[FATAL]   Executing Ansible playbook /home/ubuntu/toto/venv/lib/python3.10/site-packages/netsim/ansible/initial-config.ansible failed:
  Command '['ansible-playbook', '/home/ubuntu/toto/venv/lib/python3.10/site-packages/netsim/ansible/initial-config.ansible']' returned non-zero exit status 4.
Error executing netlab initial --no-message:
  Command '['netlab', 'initial', '--no-message']' returned non-zero exit status 1.
[FATAL]   netlab up: netlab initial failed, aborting...
```

**Comment by ipspace** [comment]
Yeah, you need https://github.com/ipspace/netlab/commit/801da692c29915851e12966c7709c515ba9f0f57.

You could either change that one line in mpls-clab.yml or I could push out 1.8.0-post2. I'll push out a new bug fix release pretty soon in any case to make sure nobody else gets bitten by this one, but it would be nice to get this one working first (in case there are other gotchas lurking in the shadows).

**Comment by barajus** [comment]
Yes, I naively thought the diff was included in `1.8.0-post1`. 
It works now  👍  Thank you!

**Comment by ipspace** [comment]
1.8.0-post2 is out. Closing this one.
